pkg/eventstream/eventstream.go

```diff
@@ -0,0 +1,98 @@
+package eventstream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+var (
+	ErrInvalidRequestBody = fmt.Errorf("Request body must contain an event object or an array of event objects")
+	ErrEventTooLarge      = fmt.Errorf("Event is over the max size")
+)
+
+// MaxEvents is the maximum number of events we can parse in a single batch.
+const MaxEvents = 1_000
+
+// ParseStream parses a reader, publishing a stream of JSON-encoded events to the given channel,
+// ensuring that no individual event is too large.
+//
+// This closes the given channel once the JSON stream in the given reader has been parsed.
+//
+// Usage:
+//
+//      var err error
+//      go func() {
+//              err = ParseStream(ctx, r, stream)
+//      ()
+//
+//      for bytes := range stream {
+//              // consume event, transform event, etc
+//      }
+//
+//      if err != nil {
+//              // handle error
+//      }
+//
+func ParseStream(ctx context.Context, r io.Reader, stream chan json.RawMessage, maxSize int) error {
+	defer func() {
+		close(stream)
+	}()
+	d := json.NewDecoder(r)
+
+	token, err := d.Token()
+	if err == io.EOF {
+		return nil
+	}
+
+	delim, ok := token.(json.Delim)
+	if !ok {
+		// Invalid type
+		return ErrInvalidRequestBody
+	}
+
+	i := 0
+
+	switch delim {
+	case '{':
+		// We've already peeked the first char.  Read all, then prepend a '{'
+		byt, err := io.ReadAll(d.Buffered())
+		if err != nil {
+			return err
+		}
+		// d.Buffered() only returns a portion of the buffered stream;  read the rest
+		// and concat.
+		extra, err := io.ReadAll(r)
+		if err != nil {
+			return err
+		}
+		data := append([]byte("{"), byt...)
+		data = append(data, extra...)
+		if len(data) > maxSize {
+			return fmt.Errorf("%w: Max %d bytes / Size %d bytes", ErrEventTooLarge, maxSize, len(data))
+		}
+
+		stream <- data
+	case '[':
+		// Parse a stream of tokens
+		for d.More() {
+			if i == MaxEvents {
+				return fmt.Errorf("maximum events parsed within a batch: %d", MaxEvents)
+			}
+
+			jsonEvt := json.RawMessage{}
+			if err := d.Decode(&jsonEvt); err != nil {
+				return err
+			}
+			if len(jsonEvt) > maxSize {
+				return fmt.Errorf("%w: Max %d bytes / Size %d bytes", ErrEventTooLarge, maxSize, len(jsonEvt))
```

Comment:
is `MaxSize` something users are aware of?
if not, might be good to have a separate issue to add it to the docs so people can be aware of it.

Reply:
also, should the error message be similar to the deleted one?

https://github.com/inngest/inngest/pull/436/files#diff-67a6e36014be865053c43d38b2091bd2d3f17a91b117d97c0f5244cd7d71127dL107